packages/lab/src/carousel/CarouselContext.tsx

```diff
@@ -0,0 +1,154 @@
+  useCallback,
+  useContext,
+  useEffect,
+  useRef,
+  useState,
+} from "react";
+
+export interface CarouselContextValue {
+  firstVisibleSlide: number;
+  visibleSlides: number;
+  slidesCount: number;
+  nextSlide: (event: MouseEvent | KeyboardEvent) => void;
+  prevSlide: (event: MouseEvent | KeyboardEvent) => void;
+  updateActiveFromScroll: (scrollLeft: number) => void;
+  focusSlide: (index: number) => void;
+  registerSlide: (element: HTMLDivElement) => number;
+  unregisterSlide: (index: number) => void;
+  containerRef: RefObject<HTMLDivElement>;
+  carouselId?: string;
+}
+
+export const CarouselContext = createContext<CarouselContextValue | null>(
+  "CarouselContext",
+  null,
+);
+
+export function useCarousel() {
+  const context = useContext(CarouselContext);
+  if (!context) {
+    throw new Error("useCarousel must be used within carousel provider");
+  }
+  return context;
+}
+
+export function CarouselProvider({
+  children,
+  activeSlideIndex = 0,
+  visibleSlides = 1,
+  id,
+}: {
+  children: ReactNode;
+  activeSlideIndex?: number;
+  visibleSlides?: number;
+  id?: string;
+}) {
+  const [firstVisibleSlide, setFirstVisibleSlide] = useState(activeSlideIndex);
+  const [visibleFocus, setVisibleFocus] = useState(0);
+  const [sliderW, setSliderW] = useState(0);
+  const slides = useRef<Map<number, HTMLDivElement>>(new Map());
+  const [slidesCount, setSlidesCount] = useState(slides.current.size);
+  useEffect(() => {
+    if (containerRef.current) {
+      scrollToSlide(activeSlideIndex);
+    }
+  }, [activeSlideIndex]);
+
+  const containerRef = useRef<HTMLDivElement>(null);
+
+  const registerSlide = useCallback((element: HTMLDivElement) => {
+    const assignedIndex = slides.current.size;
+    slides.current.set(assignedIndex, element);
+    setSlidesCount(assignedIndex + 1);
+    return assignedIndex;
+  }, []);
+
+  const unregisterSlide = useCallback((index: number) => {
+    slides.current.delete(index);
+    setSlidesCount(slides.current.size);
+  }, []);
+
+  const updateActiveFromScroll = (scrollLeft: number) => {
+    const newIndex = Math.round(scrollLeft / (sliderW / visibleSlides)) || 0;
+    if (newIndex !== firstVisibleSlide) {
+      setFirstVisibleSlide(newIndex);
+    }
+  };
+
+  const handleResize = useCallback(() => {
+    if (!containerRef.current) return;
+    setSliderW(containerRef.current.offsetWidth);
+  }, []);
+
+  useResizeObserver({ ref: containerRef, onResize: handleResize });
+
+  const scrollToSlide = (index: number) => {
+    if (!containerRef.current) return;
+    const sliderWidth = containerRef.current.offsetWidth;
+    const slideWidth = sliderWidth / visibleSlides;
+    const targetScrollLeft = index * slideWidth;
+    containerRef.current.scrollTo({
+      left: targetScrollLeft,
+      behavior: "smooth",
+    });
+  };
+
+  const focusSlide = (index: number) => {
+    slides.current.get(index)?.focus();
+  };
+
+  const nextSlide = (event: MouseEvent | KeyboardEvent) => {
```

Comment:
Next Slide and focus slide get combined here, we pass in an event, only to check it's type.
We then use that type to either give the card focus, so we can enable next/prev keyboard control or advance the card. 

I then went to useKeyNavigation and could see that we needed this code to support an internal hook. We pushed complexity into context which was created by a hook only used by CarouselNavigation.

Was not expecting this code in here, perhaps needs refactoring, so nextSlide, always advances the slide and focus is provided elsewhere?